File: docstrings/tt_header.py
# -*- coding: utf-8 -*-

# Copyright (c) 2013, Michael Droettboom All rights reserved.

# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:

# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in
#    the documentation and/or other materials provided with the
#    distribution.

# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

# The views and conclusions contained in the software and
# documentation are those of the authors and should not be interpreted
# as representing official policies, either expressed or implied, of
# the FreeBSD Project.

from __future__ import print_function, unicode_literals, absolute_import

TT_Header__init__ = """
TrueType header.
"""

TT_Header_table_version = """
The version of the TrueType header.
"""

TT_Header_font_revision = """
The version of the font.  Set by the font manufacturer.
"""

TT_Header_checksum_adjust = """
A checksum for the header.
"""

TT_Header_magic_number = """
Always 0x5F0F3CF5.
"""

TT_Header_flags = """
`TT_HEADER_FLAGS` bitflags that apply to the face as a whole.
"""

TT_HEADER_FLAGS = """
|freetypy| Bitflags giving global information about the font.

- `BASELINE_AT_ZERO`: If set, the baselines for the font is at *y* = 0
  (that is, the *x*-axis)

- `LEFT_BLACK_BIT_IS_LSB`: If set, the *x*-position of the leftmost
  black bit is assumed to be the left side bearing.

- `NONLINEAR_SCALE`: If set, instructions may use point size
  explicitly in place of pixels per em. Scaled point size and actual
  point size will differ (i.e. 24 point glyph differs from 12 point
  glyph scaled by factor of 2).

- `INTEGER_SCALING`: Use integer scaling instead of fractional.

- `VERTICAL`: Set in fonts that are intended to ne laid out
  vertically, and in which the glyphs have been drawn such that an
  *x*-coordinate of 0 corresponds to the desired vertical baseline.

- `REQUIRES_LAYOUT`: Set if the font requires layout for correct
  linguistic rendering (e.g. Arabic fonts).

- `METAMORPHOSIS_EFFECTS`: Set for a GX font which has one or more
  metamorphosis effects designated as happening by default.

- `RIGHT_TO_LEFT_GLYPHS`: Set if the font contains any strong
  right-to-left glyphs.

- `INDIC_REARRANGEMENT_EFFECTS`: Set if the font contains Indic-style
  rearrangement effects.
"""

TT_Header_units_per_em = """
Units per em.  In the range 64 - 16384.
"""

TT_Header_created = """
The date the font was created.
"""

TT_Header_modified = """
The date the font was modified.
"""

TT_Header_x_min = """
The minimum x for all bounding boxes.
"""

TT_Header_y_min = """
The minimum y for all bounding boxes.
"""

TT_Header_x_max = """
The maximum x for all bounding boxes.
"""

TT_Header_y_max = """
The maximum y for all bounding boxes.
"""

TT_Header_mac_style = """
`TT_MAC_STYLE` bitflags.
"""

TT_MAC_STYLE = """
Bit flags indicating the style of the face.

- `BOLD`

- `ITALIC`

- `UNDERLINE`

- `OUTLINE`

- `SHADOW`

- `CONDENSED`

- `EXTENDED`
"""

TT_Header_lowest_rec_ppem = """
The smallest readable size in pixels.
"""

TT_Header_font_direction = """
The direction of the font.

- 0: Mixed directional glyphs

- 1: Only strong left to right glyphs

- 2: Like 1, but also contains neutrals

- -1: Only strong right to left glyphs

- -2: Like -1, but also contains neutrals
"""

File: src/tt_header.c
/*
Copyright (c) 2015, Michael Droettboom
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this
   list of conditions and the following disclaimer.
2. Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

The views and conclusions contained in the software and documentation are those
of the authors and should not be interpreted as representing official policies,
either expressed or implied, of the FreeBSD Project.
*/

#include "tt_header.h"
#include "doc/tt_header.h"
#include "face.h"

#include FT_TRUETYPE_TABLES_H


#define MAKE_TT_HEADER_GETTER(name, convert_func, member)       \
    MAKE_GETTER(Py_TT_Header, name, convert_func, member);
#define DEF_TT_HEADER_GETTER(name) DEF_GETTER(name, doc_TT_Header_ ## name)


ftpy_ConstantType Py_TT_HEADER_FLAGS_BitflagType;
ftpy_ConstantType Py_TT_MAC_STYLE_BitflagType;


/****************************************************************************
 Object basics
*/


typedef struct {
    ftpy_Object base;
    TT_Header *x;
} Py_TT_Header;


static PyTypeObject Py_TT_Header_Type;


PyObject *
Py_TT_Header_cnew(PyObject *face)
{
    Py_Face *f = (Py_Face *)face;
    Py_TT_Header *self;
    TT_Header *orig;

    orig = (TT_Header *)FT_Get_Sfnt_Table(f->x, ft_sfnt_head);
    if (orig == NULL) {
        PyErr_SetString(PyExc_AttributeError, "No head found");
        return NULL;
    }

    self = (Py_TT_Header *)(&Py_TT_Header_Type)->tp_alloc(&Py_TT_Header_Type, 0);
    if (self == NULL) {
        return NULL;
    }
    Py_INCREF(face);
    self->base.owner = face;
    self->x = orig;
    return (PyObject *)self;
}


static int
Py_TT_Header_init(Py_TT_Header *self, PyObject *args, PyObject *kwds)
{
    PyErr_SetString(
        PyExc_RuntimeError,
        "TT_Header objects may not be instantiated directly. "
        "Use Face.tt_header instead.");
    return -1;
}


/****************************************************************************
 Getters
*/

MAKE_TT_HEADER_GETTER(table_version, ftpy_PyFloat_FromFT_FIXED, self->x->Table_Version)
MAKE_TT_HEADER_GETTER(font_revision, ftpy_PyFloat_FromFT_FIXED, self->x->Font_Revision)
MAKE_TT_HEADER_GETTER(checksum_adjust, PyLong_FromLong, self->x->CheckSum_Adjust)
MAKE_TT_HEADER_GETTER(magic_number, PyLong_FromLong, self->x->Magic_Number)


static PyObject *flags_get(Py_TT_Header *self, PyObject *closure)
{
    return Py_Constant_cnew(
        &Py_TT_HEADER_FLAGS_BitflagType, self->x->Flags);
}


MAKE_TT_HEADER_GETTER(units_per_em, PyLong_FromUnsignedLong, self->x->Units_Per_EM);
MAKE_TT_HEADER_GETTER(created, ftpy_PyDateTime_FromTTDateTime, self->x->Created);
MAKE_TT_HEADER_GETTER(modified, ftpy_PyDateTime_FromTTDateTime, self->x->Modified);
MAKE_TT_HEADER_GETTER(x_min, PyLong_FromLong, self->x->xMin);
MAKE_TT_HEADER_GETTER(y_min, PyLong_FromLong, self->x->yMin);
MAKE_TT_HEADER_GETTER(x_max, PyLong_FromLong, self->x->xMax);
MAKE_TT_HEADER_GETTER(y_max, PyLong_FromLong, self->x->yMax);


static PyObject *mac_style_get(Py_TT_Header *self, PyObject *closure)
{
    return Py_Constant_cnew(
        &Py_TT_MAC_STYLE_BitflagType, self->x->Mac_Style);
}


MAKE_TT_HEADER_GETTER(lowest_rec_ppem, PyLong_FromUnsignedLong, self->x->Lowest_Rec_PPEM);
MAKE_TT_HEADER_GETTER(font_direction, PyLong_FromLong, self->x->Font_Direction);


static PyGetSetDef Py_TT_Header_getset[] = {
    DEF_TT_HEADER_GETTER(table_version),
    DEF_TT_HEADER_GETTER(font_revision),
    DEF_TT_HEADER_GETTER(checksum_adjust),
    DEF_TT_HEADER_GETTER(magic_number),
    DEF_TT_HEADER_GETTER(flags),
    DEF_TT_HEADER_GETTER(units_per_em),
    DEF_TT_HEADER_GETTER(created),
    DEF_TT_HEADER_GETTER(modified),
    DEF_TT_HEADER_GETTER(x_min),
    DEF_TT_HEADER_GETTER(y_min),
    DEF_TT_HEADER_GETTER(x_max),
    DEF_TT_HEADER_GETTER(y_max),
    DEF_TT_HEADER_GETTER(mac_style),
    DEF_TT_HEADER_GETTER(lowest_rec_ppem),
    DEF_TT_HEADER_GETTER(font_direction),
    {NULL}  /* Sentinel */
};


/****************************************************************************
 Setup
*/


static PyTypeObject Py_TT_Header_Type;


static PyTypeObject Py_TT_HEADER_FLAGS_Type;
#define TT_HEADER_FLAGS_CONST(name) DEF_CONST(name, TT_HEADER_FLAGS)
static constant_def TT_HEADER_FLAGS_constants[] = {
    {"BASELINE_AT_ZERO", 0x1},
    {"LEFT_BLACK_BIT_IS_LSB", 0x2},
    {"NONLINEAR_SCALE", 0x4},
    {"INTEGER_SCALING", 0x8},
    {"VERTICAL", 0x20},
    {"REQUIRES_LAYOUT", 0x80},
    {"METAMORPHOSIS_EFFECTS", 0x100},
    {"RIGHT_TO_LEFT_GLYPHS", 0x200},
    {"INDIC_REARRANGEMENT_EFFECTS", 0x400},
    {NULL}
};


static PyTypeObject Py_TT_MAC_STYLE_Type;
#define TT_MAC_STYLE_CONST(name) DEF_CONST(name, TT_MAC_STYLE)
static constant_def TT_MAC_STYLE_constants[] = {
    {"BOLD", 0x1},
    {"ITALIC", 0x2},
    {"UNDERLINE", 0x4},
    {"OUTLINE", 0x8},
    {"SHADOW", 0x10},
    {"CONDENSED", 0x20},
    {"EXTENDED", 0x40},
    {NULL}
};


int setup_TT_Header(PyObject *m)
{
    memset(&Py_TT_Header_Type, 0, sizeof(PyTypeObject));
    Py_TT_Header_Type = (PyTypeObject) {
        .tp_name = "freetypy.TT_Header",
        .tp_basicsize = sizeof(Py_TT_Header),
        .tp_doc = doc_TT_Header__init__,
        .tp_getset = Py_TT_Header_getset,
        .tp_init = (initproc)Py_TT_Header_init,
    };

    ftpy_setup_type(m, &Py_TT_Header_Type);

    if (define_bitflag_namespace(
            m, &Py_TT_HEADER_FLAGS_Type, &Py_TT_HEADER_FLAGS_BitflagType,
            "freetypy.TT_HEADER_FLAGS",
            doc_TT_HEADER_FLAGS, TT_HEADER_FLAGS_constants)) {
        return -1;
    }

    if (define_bitflag_namespace(
            m, &Py_TT_MAC_STYLE_Type, &Py_TT_MAC_STYLE_BitflagType,
            "freetypy.TT_MAC_STYLE",
            doc_TT_MAC_STYLE, TT_MAC_STYLE_constants)) {
        return -1;
    }

    return 0;
}
